#7-User Authentication

mernxl opened the pull request:
## Description
Implementation of authentication, accessible through `/auth/{action}` where **action** stands for the action to be carried out. These actions include;

- **sign-up**: Normal signup, takes at least an email and password 
- **sign-in**: Sign in with an email and password
- **sign-out**: Clear the session information
- **update-password**: Update password of current user
- **reset-password**: Reset password using a token generated and sent via mail or other methods
- **request-reset-password**: Request the reset Token for password reset.

### Other Notes
- Passwords are stored in a schema on the user and by default are never populated when requesting user. The password value is harshed using `bcrypt`.
- On many failed login attempts, we lock user account for a given period.

You can check up my [comment on issue](https://github.com/BuildForSDG/voxnostra/issues/7#issuecomment-626113129) for more

Fixes #7 

## How Has This Been Tested?
I did minor testing with PostMan to see that the features. Still to in app tests

## Checklist:
<!--- Put an `x` in all the boxes that apply ! -->
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added necessary inline code documentation
- [ ] I have added tests that prove my fix is effective and that this feature works
- [ ] New and existing unit tests pass locally with my changes

mernxl committed:
feat: add api error classes, implement catch all handlers in middlewares

mernxl committed:
feat(middlewares): add `json` method to `res` object

mernxl committed:
chore(deps): add `bcrypt`, `jsonwebtoken`, and `moment` packages

mernxl committed:
chore(props): add Authentication dependent properties to voxnostra.properties file

mernxl committed:
feat(libs): add authentication helpers, configure constants

mernxl committed:
fix(libs): fix db initialisation logic

mernxl committed:
fix(libs): fix session setup logic

mernxl committed:
refactor(middleware): add console.error logging for caught error

mernxl committed:
feat(libs): add `setUserSession` function to auth-helpers, fix config constant init

mernxl committed:
feat(schema): add password embedded schema, move user.model into user folder

mernxl committed:
feat(schema): add password embedded schema, move user.model into user folder

mernxl committed:
feat(api): add user-authentication api endpoints

mernxl committed:
7-feat(api): swap testapi endpoint with a health-check endpoint

tanerochris commented on middlewares/index.js:
```diff
@@ -6,20 +6,50 @@ const cors = ({ req, res }) => {
+
+
+  res.statusCode = 404;
+  return res.end('Resource Not Found');
```
this line returns a 404 status even when the request was successful. I think these lines should be removed.

tanerochris commented on pages/api/auth/sign-up.js:
```diff
@@ -0,0 +1,38 @@
+
+    setUserSession(req, user);
+
+    res.json(res.session.user);
```
this should be req.session.user

mernxl replied:
Ok, they are currently temporary return values. But will fix it now.

tanerochris commented on schemas/user/user.model.js:
```diff
@@ -0,0 +1,18 @@
+
+const UserSchema = new Schema({
+  name: String,
+  email: {
+    type: String, index: true, unique: true, required: true
+  },
```
enforce email address validation, but can be done in next commit.

mernxl replied:
I will just do that now.

tanerochris commented:
@mernxl fix the codacy errors

tanerochris committed:
6-docs(): Api design documentation
OpenAPI 3.0 design
Yaml and JSON files

closes #6

mernxl commented:
![image](https://user-images.githubusercontent.com/16241595/81540304-5aa78a80-9369-11ea-8822-61f45c73e6aa.png)

The review from codacity is somewhat incorrect, well will raise an issue later. Clearly i destructured that object which is valid in JS. Well i will fix it now, raise a flag in codacity.com so we move on.

mernxl committed:
7-feat(schema): add email validation regex with onFail message

mernxl committed:
7-fix(middleware): fix sending 404 after response sent

- set content type for catchall error response

mernxl committed:
7-fix(api): fix res.session is undefined for reset-password

- workaround for JS Object destructuring inside an if block considered error by codacity

tanerochris reviewed (approved):
changes can be merged

tanerochris committed:
6-docs(enpoints): updated API docs readme
Added instructions to run API endpoint documentation
Added link to external docs

close

tanerochris commented on schemas/user/password/password.schema.js:
```diff
@@ -0,0 +1,47 @@
+      // generate a salt
+      const salt = await bcrypt.genSalt(SALT_WORK_FACTOR);
+
+      // replacing clear-text with harsh
+      // harsh the password along with our new salt
+      this.value = await bcrypt.hash(this.value, salt);
```
this.value is undefined, produces the error data and salt arguments required.

mernxl committed:
7-fix(api): fix error when password is null on create

KaiserPhemi commented on schemas/user/user.model.js:
```diff
@@ -0,0 +1,22 @@
+    index: true,
+    unique: true,
+    required: true,
+    validate: [/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/, 'Must be a valid email address.']
```
# Nitpick
This makes the code a bit difficult to read. Why not abstract the data validation to another function or library like [Joi](https://hapi.dev/module/joi/)

mernxl replied:
Ok, I was to raise the issue of validation. A framework was supposed to be set before i worked on this task. That is to permit us carry out validations at the controller level.

tanerochris replied:
Can we leave validation currently at the database, validation at controller will be functionality update on another issue.

mernxl commented:
Well fixed it. 😅In case future such mistakes.

KaiserPhemi commented on libs/api-errors.js:
```diff
@@ -0,0 +1,42 @@
+ *
+ * name, message, stack
+ */
+export class ErrorResponse extends Error {
```
# Nitpick
I understand that this handles error messages for the API but if we can include all that in the response object in the controller, there won't be a need for this.

mernxl replied:
@KaiserPhemi you mean like, i create an error method on the response object, like i did for .json to send json responses??

KaiserPhemi replied:
@mernxl 
Typically you'll send an API response like this:
```js
return response.send({
message: "",
data,
})
```

This class appears to only handle error message. What I'm saying is that, why create a class for that when I can easily send the error message as an API response like this: 

```js
return response
  .status(500)
  .send({
     message: "Internal Server Error",
     details,
});
```

mernxl replied:
Ok i see. Maybe i state the reasons for my doing that, you see if they are legit.

Although i know theres a down side of having to deal with the error generation and handling but the reasons for doing that are as follows.

- I wanted to reuse the error messages, status codes, so i don't have to always create new ones which may mismatch.
- Gives a general and continuous structure for all backend errors. Especially when it comes to validation errors.

KaiserPhemi commented on libs/mongoose/Database.js:
```diff
@@ -5,12 +5,19 @@ import PropertiesReader from 'properties-reader';
+    await connect(mongooseUrl,
+      { useNewUrlParser: true, useCreateIndex: true, useUnifiedTopology: true });
+    isFirst = false;
+    // fixme: here we log connection errors and retries
```
# Nitpick
Remove commented codes

mernxl replied:
Most of what i commented are task that still need to be done and not related to my task, Like there is no logging infrastructure setup for the code base.

Once that task is done, i will add it to the code base, I should probably create an issue for that.

KaiserPhemi reviewed (changes requested):
# Critical
- The folder structure seems a bit cumbersome. This does not make it easy for other users to navigate through the project during development
- Many files aren't documented properly i.e. functions, classes & methods should carry at least the following docs:
```js
/**
* @desc description of what the function does
* @param {type} param
*/
```

- All database related files or functions should be in a models folder i.e. `/models` not `/libs`
- Also, all frontend related files should be in a separate folder for easy identification cos as it is, everything seems to be not well structured.

- Please attach screenshots of test results.

mernxl commented:
@KaiserPhemi Thank you for your reviews. I guess the whole structure will need to be changes in another pull request then i rebase and add this changes to follow up a better structure. That was actually the structure i found so i had to flow with it.

@tanerochris maybe we will have to look into the file structure, given you setup the initial one.

- About the code documenting, i will document all the functions i created for this tasks.
- For tests, i will merge the test feature

mernxl commented:
I noticed @tanerochris already took in the changes of this pull request in #20 pull. I guess this pull request will see a closure then, So @tanerochris will have to do the updates on this pull if thats ok.

tanerochris commented:
I think having an error response structure is appropriate, especially for
writing tests, also it prevents a scenario where other persons create
different error responses. In #20 I added some sort of adapter to the
ErrorResponse classes.

On Mon, May 18, 2020 at 1:00 PM Rodrick Nfinyoh M <notifications@github.com>
wrote:

> *@mernxl* commented on this pull request.
> ------------------------------
>
> In libs/api-errors.js
> <https://github.com/BuildForSDG/voxnostra/pull/16#discussion_r426862527>:
>
> > @@ -0,0 +1,42 @@
> +/* eslint-disable max-classes-per-file */
> +
> +/**
> + * General Error Object, gets other props from Error which include
> + *
> + * name, message, stack
> + */
> +export class ErrorResponse extends Error {
>
> Ok i see. Maybe i state the reasons for my doing that, you see if they are
> legit.
>
> Although i know theres a down side of having to deal with the error
> generation and handling but the reasons for doing that are as follows.
>
>    - I wanted to reuse the error messages, status codes, so i don't have
>    to always create new ones which may mismatch.
>    - Gives a general and continuous structure for all backend errors.
>    Especially when it comes to validation errors.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BuildForSDG/voxnostra/pull/16#discussion_r426862527>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACC3JXPMVM6GB6PJLUQZWQDRSGHWBANCNFSM4M47L5YA>
> .
>


-- 
My twitter  @tanerochris
What I've built https://tchizer.com
Developer Circle Buea Lead facebook.com/groups/DevCBuea/
Whatsapp +237 678431904

mernxl committed:
4-feat(schema): add app schema based on api-docs and endpoints

- add project schema
- add user schema

mernxl committed:
7-fix(middleware): fix sending 404 after response sent

mernxl committed:
7-feat(libs): add isAuthenticated method in auth-helpers, add index file in libs dir

mernxl committed:
7-docs(schema): add documentation to a password schema method

mernxl committed:
7-feat(api): add /user/me path, refactor routes to suit api-docs
